File: FirefighterStats/Server/Entities/Firefighter.cs
// -----------------------------------------------------------------------
//  <copyright project="FirefighterStats.Server" file="Firefighter.cs" company="syuko">
//  Copyright (c) syuko. All rights reserved.
//  </copyright>
// -----------------------------------------------------------------------

namespace FirefighterStats.Server.Entities;

using FirefighterStats.Shared.Firefighter;
using JetBrains.Annotations;
using Microsoft.AspNetCore.Identity;

[UsedImplicitly]
public class Firefighter : IdentityUser
{
    public DateTime? CareerStartDate { get; set; }

    public string? FireStation { get; set; }

    public required string FirstName { get; set; }

    public IEnumerable<IndemnitySlip>? IndemnitySlips { get; set; }

    public required string LastName { get; set; }

    public EFirefighterRank? Rank { get; set; }

    public string? RegistrationNumber { get; set; }
}

File: FirefighterStats/Client/src/types/Firefighter.ts
import EFirefighterRank from './EFirefighterRank';

type Firefighter = {
    firstName: string;
    lastName: string;
    rank?: keyof typeof EFirefighterRank | null;
    careerStartDate?: string | null;
    fireStation?: string | null;
    registrationNumber?: string | null;
};

export function isFirefighter(data: any): data is Firefighter {
    return (
        typeof data.firstName === 'string' &&
        typeof data.lastName === 'string' &&
        (Object.keys(EFirefighterRank).includes(data.rank) || data.rank === null) &&
        (typeof data.careerStartDate === 'string' || data.careerStartDate === null) &&
        (typeof data.fireStation === 'string' || data.fireStation === null) &&
        (typeof data.registrationNumber === 'string' || data.registrationNumber === null)
    );
}

export default Firefighter;
